### PR TITLE
Adding files for the NewsApi challenge

### PR DESCRIPTION
**Pull request by AineKiraboMbabazi**
#NewApi Challenge
-User is offered a list of four news sources.
- User should be able to make a choice of a preferred news source. 
-User gets back a list of the top 10 headlines from the preferred source.
-
screenshot from 2018-09-17 21-08-41
screenshot from 2018-09-17 21-09-05
screenshot from 2018-09-18 15-06-22

 A news headline should have a title, description and a url in case the user needs to follow up on the news story.

### CODE REVIEW

**Comment by AineKiraboMbabazi** [comment]
@dennisja 
@janetnamutebi
@malmike
please Review my solution for the newsapi challenge

**Comment by dennisja** [comment]
Remove all files in the `.pytest_cache` folder from version control

File: apicode/newapi.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    def sort_and_display_results(self):
+        content = self.store_api_response()
+        article_count = 1
+        print('articles length', len(content))
+        print("\t", '{:=^100}'.format('.'))
+        print("\t", '{: ^100}'.format('Top 10 Headlines from '))
+        print("\t", '{:=^100}'.format('.'))
+        for article in content:
+            print("\t", '{:_^100}'.format(article_count))
```
Comment (dennisja) [inline apicode/newapi.py]: Would appreciate if you could just build a string and print it out using one print function

File: apicode/newapi.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        print(answers['sources'])
+
+        return answers['sources']
+    # function to retrieve a product key for the environment variable
+
+    def add_api_key(self):
+        if('API_KEY' in os.environ) == False:
+            raise ValueError(
+                "Access Denied ..., you dont have a product key in your environment variables")
+        return os.environ.get('API_KEY')
```
Comment (dennisja) [inline apicode/newapi.py]: Good attempt of getting the apiKey from environment variables

File: apicode/newapi.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        ]
+        # theme adds a green background to the items in the list
+        answers = inquirer.prompt(questions, theme=GreenPassion())
+        print(answers['sources'])
+
+        return answers['sources']
+    # function to retrieve a product key for the environment variable
+
+    def add_api_key(self):
```
Comment (dennisja) [inline apicode/newapi.py]: This method name is annoying. It would be better if you could name it `get_api_key`

File: apicode/newapi.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        if('API_KEY' in os.environ) == False:
+            raise ValueError(
+                "Access Denied ..., you dont have a product key in your environment variables")
+        return os.environ.get('API_KEY')
+
+    # function for collecting data for from the API
+
+    def fetch_data(self, newssource):
+
+        api_key = self.add_api_key()
```
Comment (dennisja) [inline apicode/newapi.py]: The api key method raises an exception.
When using such a method, you should wrap it in a `try...except` block to catch any errors that might occur

Comment (dennisja) [reply apicode/newapi.py]: This helps prevent your app from crashing

File: apicode/newapi.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        api_response = self.fetch_data(self.get_user_source())
+        response_data = api_response.json()
+        headlineArticles = response_data['articles']
+
+        return headlineArticles
+
+    def sort_and_display_results(self):
+        content = self.store_api_response()
+        article_count = 1
+        print('articles length', len(content))
```
Comment (dennisja) [inline apicode/newapi.py]: Build a string and print it using 1 print statement

File: apicode/newapi.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    # function for storing the response for manipulation
+
+    def store_api_response(self):
+        api_response = self.fetch_data(self.get_user_source())
+        response_data = api_response.json()
+        headlineArticles = response_data['articles']
+
+        return headlineArticles
+
+    def sort_and_display_results(self):
```
Comment (dennisja) [inline apicode/newapi.py]: This method is called `sort_and_display_results` but, I don't see any where you are sorting the results from

**Review by dennisja — CHANGES_REQUESTED** [review]

File: apicode/newapi.py (hunk 1 above)
Comment (AineKiraboMbabazi) [reply apicode/newapi.py]: I have addressed this issue by using a texttable module in python

File: apicode/newapi.py (hunk 2 above)
Comment (AineKiraboMbabazi) [reply apicode/newapi.py]: thank you, it was Malmikes idea.

File: apicode/newapi.py (hunk 3 above)
Comment (AineKiraboMbabazi) [reply apicode/newapi.py]: Hahaha to be changed immediately

File: apicode/newapi.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+        return headlineArticles, selectedSource
+
+    def display_results(self):
+
+        content = self.store_api_response()
+        article_count = 1
+        source_selected = content[1]
+
+        print('Headlines retrieved : ', len(content[0]))
```
Comment (dennisja) [inline apicode/newapi.py]: Build this using a single string

File: apicode/newapi.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        ]
+        # theme adds a green background to the items in the list
+        answers = inquirer.prompt(questions, theme=GreenPassion())
+        print("News Source : ", answers['sources'])
+
+        return answers['sources']
+    # function to retrieve a product key for the environment variable
+
+    def get_api_key(self):
+        try:
```
Comment (dennisja) [inline apicode/newapi.py]: This is unneccessary. The code in except will never be executed, The get method returns none if no key is found

File: apicode/newapi.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+            os.environ.get('API_KEY')
+        except Exception:
+
+            if 'API_KEY' not in os.environ:
+                print(
```
Comment (dennisja) [inline apicode/newapi.py]: Would be better to raise an exception, to show that the api key is missing

**Review by dennisja — CHANGES_REQUESTED** [review]

**Comment by dennisja** [comment]
Include the dependencies of your application in a `requirements` file
Update your readme to indicate how a person should set up your application and run tests and use it

File: README.md [hunk 10]
```diff
@@ -1,2 +1,60 @@
 # NewsAPI
 User is offered a list of four news sources. User should be able to make a choice of a preferred news source. User gets back a list of the top 10 headlines from the preferred source. A news headline should have a title, description and a url in case the user needs to follow up on the news story.
+
+# Requirements
```
Comment (dennisja) [inline README.md]: These shouldn't be ~~packages~~ dependencies cuz they are there in the requirements file. These should be the things one should have inorder to setup your project

File: README.md [hunk 11]
```diff
@@ -1,2 +1,60 @@
...
+
+    inquirer==2.4.0
+    requests==2.19.1
+    texttable==1.4.0
+
+# Installation and settup
+Clone the repository
+
+$ git clone https://github.com/AineKiraboMbabazi/NewsAPI.git
```
Comment (dennisja) [inline README.md]: Please make this a code block. 
Make all other commands code blocks

File: apicode/newapi.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    def display_results(self):
+
+        content = self.store_api_response()
+        article_count = 1
+        source_selected = str(content[1])
+        modifier = "="*110
+        numberofitems = str(len(content[0]))
+        title = "TOP 10 HEADLINES FROM : "
+        #creating the header for the display table.
+        print('%s %s \n %s \n \t\t\t %s %s \n %s' % ('Headlines retrieved : ',
```
Comment (dennisja) [inline apicode/newapi.py]: Using `.format` is better. You can look into how you can use it [here](https://pyformat.info/)

And if you are using python 3.6+
`f strings` are the best. You can look at them [here](https://www.python.org/dev/peps/pep-0498/?)

File: README.md (hunk 11 above)
Comment (dennisja) [reply README.md]: Read this [guide to write a better readme](https://guides.github.com/features/mastering-markdown/)

**Review by dennisja — CHANGES_REQUESTED** [review]

File: apicode/newapi.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+    def fetch_data(self, newssource):
+
+        api_key = self.get_api_key()
+
+        url = 'https://newsapi.org/v2/top-headlines?sources='+newssource + \
+            '&apiKey='+api_key+'&pageSize=10'
+
+        return (requests.get(url))
+    # function for checking whether the api call was successfull
```
Comment (dennisja) [inline apicode/newapi.py]: Make these comments doc strings and put them at the start of the function e.g.
```python
# good
def my_awesome_method():
     """ Does awesome things """
     do_something_awesome()

# bad
# does awesome things
def my_awesome_method():
   do_something_awesome()
```

**Review by dennisja — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (AineKiraboMbabazi) [reply README.md]: I am having trouble understanding this comment, did you mean that i should put things like the user should have python 3.6 installed?

Comment (dennisja) [reply README.md]: You are specifying your project dependencies yet they are already in the requirements.txt file. The requirements here should be things one should have set up to run your project e.g One should have PythonXX, e.t.c

File: .travis.yml [hunk 14]
```diff
@@ -0,0 +1,11 @@
...
+python:
+  
+  - "3.6"
+ 
+# command to install dependencies
+install:
+  - pip install -r requirements.txt
+# command to run tests
+script:
+  - pytest
```
Comment (dennisja) [inline .travis.yml]: Point pytest to the folder containing your application ~~code~~ tests to fix the Travis build
```yml
script:
  - pytest tests
```

**Review by dennisja — CHANGES_REQUESTED** [review]

File: apicode/newapi.py (hunk 13 above)
Comment (AineKiraboMbabazi) [reply apicode/newapi.py]: noted

File: .travis.yml (hunk 14 above)
Comment (dennisja) [reply .travis.yml]: @AineKiraboMbabazi 👆 I have updated this. Sorry for the initial info. I was on rush. The first comment applies to coverage

**Comment by dennisja** [comment]
Good Job on running tests.
Can you do the same on coverage?
look at how you can leverage https://pypi.org/project/pytest-cov/ and coveralls for test reporting

**Comment by AineKiraboMbabazi** [comment]
Thank you.
Am getting started on it right now

File: .coveralls.yml [hunk 15]
```diff
@@ -0,0 +1 @@
+repo_token: zbmrHOitciQcFigB3CwnHTpTpqp1gn0lV
```
Comment (dennisja) [inline .coveralls.yml]: What is the repo token doing here?

File: .coverage [hunk 16]
```diff
@@ -0,0 +1 @@
+!coverage.py: This is a private format, don't read it directly!{"lines":{"/home/andela/Projects/cli_tests/app/__init__.py":[1],"/home/andela/Projects/cli_tests/app/app_Api.py":[1,2,3,4,5,8,9,11,28,36,49,61,70,106,63,38,41,42,30,31,34,45,47,64,66,68,52,53,54,55,56,58,73,74,75,76,77,78,80,81,84,85,86,87,89,90,91,92,93,94,95,96,97,98,99,101,102,103]}}
```
Comment (dennisja) [inline .coverage:1]: The .coverage file shouldn't be in version control

File: .travis.yml [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+ 
+# command to install dependencies
+install:
+  - pip install -r requirements.txt
+  - pip install coverage
+  - pip install coveralls
+  
+# command to run tests
+script: 
+  - pytest tests
```
Comment (dennisja) [inline .travis.yml]: This command doesn't get coverage

**Review by dennisja — CHANGES_REQUESTED** [review]

File: .travis.yml [hunk 18]
```diff
@@ -6,12 +6,12 @@ python:
 # command to install dependencies
 install:
   - pip install -r requirements.txt
-  - pip install coverage
+  - pip install pytest-cov
```
Comment (dennisja) [inline .travis.yml]: This should be in your requirements.txt

File: .travis.yml [hunk 19]
```diff
@@ -6,12 +6,12 @@ python:
...
   - pip install coveralls
   
 # command to run tests
 script: 
-  - pytest tests 
+  - pytest --cov=app/
   - coveralls
```
Comment (dennisja) [inline .travis.yml]: You are running this command twice.
Here and in the after scripts block

**Pull request merged**